File: src/main/java/org/distril/beengine/command/data/ArgumentType.java
package org.distril.beengine.command.data;

import com.nukkitx.protocol.bedrock.data.command.CommandEnumData;
import com.nukkitx.protocol.bedrock.data.command.CommandParam;
import lombok.experimental.UtilityClass;
import org.distril.beengine.command.parser.StringParser;
import org.distril.beengine.command.parser.TargetParser;

@UtilityClass
public class ArgumentType {

	public static CommandArgument Integer(String name) {
		return ArgumentType.Integer(name, false);
	}

	public static CommandArgument Integer(String name, boolean optional) {
		return new CommandArgument(name, CommandParam.INT, optional);
	}

	public static CommandArgument Float(String name) {
		return ArgumentType.Float(name, false);
	}

	public static CommandArgument Float(String name, boolean optional) {
		return new CommandArgument(name, CommandParam.FLOAT, optional);
	}

	public static CommandArgument Target(String name) {
		return ArgumentType.Target(name, false);
	}

	public static CommandArgument Target(String name, boolean optional) {
		return new CommandArgument(name, CommandParam.TARGET, TargetParser.INSTANCE, optional);
	}

	public static CommandArgument String(String name) {
		return ArgumentType.String(name, false);
	}

	public static CommandArgument String(String name, boolean optional) {
		return new CommandArgument(name, CommandParam.STRING, optional);
	}

	public static CommandArgument Enum(String name, String... values) {
		return ArgumentType.Enum(name, false, values);
	}

	public static CommandArgument Enum(String name, boolean optional, String... values) {
		return new CommandArgument(name, optional, values);
	}

	public static CommandArgument GameMode(String name) {
		return ArgumentType.GameMode(name, false);
	}

	public static CommandArgument GameMode(String name, boolean optional) {
		return ArgumentType.MinecraftEnum(name, "GameMode", optional);
	}

	private static CommandArgument MinecraftEnum(String name, String enumType, boolean optional) {
		return new CommandArgument(name, CommandParam.TEXT, optional, StringParser.INSTANCE, new CommandEnumData(enumType,
				new String[0], false));
	}
}

File: src/main/java/org/distril/beengine/command/data/ArgumentType.kt
package org.distril.beengine.command.data

import com.nukkitx.protocol.bedrock.data.command.CommandEnumData
import com.nukkitx.protocol.bedrock.data.command.CommandParam
import org.distril.beengine.command.parser.EnumParser
import org.distril.beengine.command.parser.TargetParser

object ArgumentType {

	fun integer(name: String, optional: Boolean = false) = CommandArgument(name, CommandParam.INT, optional)

	fun float(name: String, optional: Boolean = false) = CommandArgument(name, CommandParam.FLOAT, optional)

	fun target(name: String, optional: Boolean = false) =
		CommandArgument(name, CommandParam.TARGET, optional, TargetParser)

	fun string(name: String, optional: Boolean = false) = CommandArgument(name, CommandParam.STRING, optional)

	fun enum(name: String, vararg values: String, optional: Boolean = false): CommandArgument {
		return CommandArgument(name, optional, *values)
	}

	fun gameMode(name: String, optional: Boolean = false) = this.minecraftEnum(name, "GameMode", optional)

	fun minecraftEnum(name: String, enumType: String, optional: Boolean): CommandArgument {
		return CommandArgument(
			name, CommandParam.TEXT, optional, EnumParser,
			CommandEnumData(enumType, Array(0) { "" }, false)
		)
	}
}
